src/Foundatio/Caching/InMemoryCacheClient.cs

```diff
@@ -189,26 +195,33 @@ public Task<int> RemoveByPrefixAsync(string prefix)
 
     internal void RemoveExpiredKey(string key, bool sendNotification = true)
     {
-        _logger.LogDebug("Removing expired cache entry {Key}", key);
+        // Consideration: We could reduce the amount of calls to this by updating ExpiresAt and only having maintenance remove keys.
+        if (_memory.TryGetValue(key, out var existingEntry) && existingEntry.ExpiresAt < SystemClock.UtcNow)
+        {
+            if (_memory.TryRemove(key, out var removedEntry))
+            {
+                if (removedEntry.ExpiresAt >= SystemClock.UtcNow)
+                    throw new Exception("Removed item was not expired");
```

Comment:
This was actively removing expired items because this method wouldn't get inlined / executed synchronously all the time when this method was called. So concurrent Add/Set calls could update the expired value causing issues. I think this could be better reworked to have a single point for all removes.